readium/shared/src/main/java/org/readium/r2/shared/util/Url.kt

```diff
@@ -173,9 +173,39 @@ public sealed class Url : Parcelable {
     public open fun relativize(url: Url): Url =
         checkNotNull(toURI().relativize(url.toURI()).toUrl())
 
+    /**
+     * Normalizes the URL using a subset of the RFC-3986 rules.
+     *
+     * https://datatracker.ietf.org/doc/html/rfc3986#section-6
+     */
+    public open fun normalize(): Url =
+        uri.buildUpon()
+            .apply {
+                path?.let {
+                    var normalizedPath = File(it).normalize().path
+                    if (it.endsWith("/")) {
+                        normalizedPath += "/"
+                    }
+                    path(normalizedPath)
+                }
+
+                if (this@Url is AbsoluteUrl) {
+                    scheme(scheme.value)
+                }
+            }
+            .build()
+            .toUrl()!!
+
     override fun toString(): String =
         uri.toString()
 
+    /**
+     * Returns whether two URLs are strictly equal, by comparing their string representation.
+     *
+     * WARNING: Strict URL comparisons can be a source of bug, if the URLs are not normalized.
+     * In most cases, you should compare using [Url.isEquivalent].
+     */
+    @DelicateReadiumApi
     override fun equals(other: Any?): Boolean {
```

Comment:
Textual `equals` makes perfect sense to me. The only places where I think we _must_ use equivalence are internal data provided referenced by an URL (`LcpDecryptor` and `EpubDeobsfucator`):
- the exchange form of URLs is the one you can find in manifest (or OPF, or Zip entry)
- URLs that apps provide to the toolkit are supposed to come from the readingOrder (what else?), which means they are in the original form.

Reply:
In practice I didn't find any place where it made sense to use a strict comparison, except for when you want to check for `null`. Adding `@DelicateReadiumApi` helped me spot places I missed and is appropriate in my opinion. This is just a warning to make you pause to think if this is really what you want to use.

Reply:
That's because you've been using equivalence as a default. I would have used it only when strictly necessary, that is, as I pointed, very rarely. It doesn't matter though.